### ISSUE TITLE
Improve deletion ergonomics

### ISSUE THREAD

**Issue by JohnDoneth** [msiglreith/grr#2]
I assume `Copy`  is not implemented because ownership is tied to that specific number but I'm struggling to figure out how to make the following code more ergonomic. I've resorted to using `Option` in several places to mitigate the borrow checker using `Option::take()`.

```rust
pub struct UploadedData {
    pub grr: Rc<grr::Device>,
    pub buffer: grr::Buffer,
    pub length: usize,
}

impl Drop for UploadedData {
    fn drop(&mut self) {
        // cannot move `self.buffer` out of borrowed content 
        // as `grr::delete_buffer` requires it's argument to be "owned".
        self.grr.delete_buffer(self.buffer); 
    }
}
```
If this were raw OpenGL I could simply free the buffer in the `Drop` impl using the `GLuint` handle hidden inside `grr::Buffer`.

Edit:
Just thinking out loud here but maybe an unsafe alternative that takes the value by reference would work.

**Comment by msiglreith** [comment]
Hi and sry, I haven't enabled notifications for this project so I haven't noticed your issue.

I can see a few different resolutions for this right now:

* Add a `Copy` feature which allows resources to be copyable. The only issue I have with this is that features might not propagate that well through middleware (?). For example grr-imgui needs to reexpose the feature then I guess
* Make every resource `Copy`

Regarding `grr::Device`:
* global: Drop the `Device` with the struct generator and move towards a [global generator](https://github.com/brendanzab/gl-rs/tree/master/gl_generator#global-generator). Every function would then be simply `grr_draw_indexed` for example (almost like in Mantle..). Need to check compatibility with the multicontext sample

There could also be an mixture of these mentioned approaches:
* Copy + global: `grr_copy_image(src_image, dst_image);`,
* 'vulkan': Copy + global + device: `grr_copy_image(device, src_image, dst_image)`

Thoughts 😃 ?

**Comment by msiglreith** [comment]
The last 'vulkan' approach would probably be the closest to a 'redesign' of OpenGL with a more modern API

**Comment by msiglreith** [comment]
Addendum: As there is an similar 'issue' for the gfx hal API: In gfx-ocean I went with explicit destructors (https://github.com/msiglreith/gfx_ocean/blob/master/src/ocean.rs#L111)

**Comment by JohnDoneth** [comment]
I like the idea of making resources `Copy`. I'm not too sure about how it would affect middlewares as you say. I haven't used any middlewares with `grr` yet to see how it would affect the API surface. 

For `Device` I haven't thought about the possibility of removing it. Is there a benefit to keeping the `Device` around? Probably thread safety by removing `Send + Sync` from `Device`? I'm assuming we can't use `Device` from multiple threads due to OpenGL's singlethreaded design, correct me if I'm wrong.

I'll take a closer look when I get to a real computer (Currently on my phone).

**Comment by msiglreith** [comment]
>  I'm not too sure about how it would affect middlewares as you say. 

My concerns are mostly for making it as an optional feature. There might be then two different grr versions with conflicting versions (application vs middleware).

> Is there a benefit to keeping the Device around? Probably thread safety by removing Send + Sync from Device? I'm assuming we can't use Device from multiple threads due to OpenGL's singlethreaded design, correct me if I'm wrong.

Device mostly correlates to to one opengl context as it provides the function entry points. In this regard it is similar to Vulkan's concept of device or instance. Beside the requirement of making the context current, OpenGL still supports multithreading ok-ish.

I'm currently leaning towards the vulkan approach ala `grr::draw(device, 3, 1, 4, 0)`. It doesn't impose an API restrictions on the user as all objects are `Copy`. On the other hand explicit multicontext support is still fine without relaying on drivers or wrappers exposing the same function pointers.

**Comment by msiglreith** [comment]
Current roadmap proposal would be:
* Make objects `Copy` except Device
* Mark the functions as `unsafe` as they actually are..
* Investigate the future of `Device`. As the internal function pointer struct is huge, it would to be transformed into a handle (ptr to a Box). I see it mostly as ergonomic tradeoff here (safe-opionated-annoying vs unsafe-simple)

**Comment by JohnDoneth** [comment]
LGTM! Thanks for the explanations. I think the Vulkan-esque approach works well. I think that marking functions as `unsafe` would be a welcome change. I've crashed a couple of times in my otherwise "safe" code due to invalid usage.

**Comment by msiglreith** [comment]
First two points should have landed on master.
Will make some additional cleanups and changes before releasing a new version on crates.io at the end of the week.

**Comment by msiglreith** [comment]
Started some experiments on the `handle` branch to see how a more unopinionated/vulkanic API checks out: https://github.com/msiglreith/grr/blob/handle/examples/handle.rs#L53-L160

`handle`
```rust
let device = grr::create_device(
    |symbol| window.get_proc_address(symbol) as *const _,
    grr::Debug::Enable {
        callback: |report, _, _, _, msg| {
            println!("{:?}: {:?}", report, msg);
        },
        flags: grr::DebugReport::FULL,
    },
);

let vs = grr::create_shader(device, grr::ShaderStage::Vertex, VERTEX_SRC.as_bytes())?;
let fs = grr::create_shader(device, grr::ShaderStage::Fragment, FRAGMENT_SRC.as_bytes())?;

let pipeline = grr::create_graphics_pipeline(
    device,
    grr::VertexPipelineDesc {
        vertex_shader: vs,
        tessellation_control_shader: None,
        tessellation_evaluation_shader: None,
        geometry_shader: None,
        fragment_shader: Some(fs),
    },
)?;
```

`master`
```rust
let grr = grr::Device::new(
    |symbol| window.get_proc_address(symbol) as *const _,
    grr::Debug::Enable {
        callback: |report, _, _, _, msg| {
            println!("{:?}: {:?}", report, msg);
        },
        flags: grr::DebugReport::FULL,
    },
);

let vs = grr.create_shader(grr::ShaderStage::Vertex, VERTEX_SRC.as_bytes())?;
let fs = grr.create_shader(grr::ShaderStage::Fragment, FRAGMENT_SRC.as_bytes())?;

let pipeline = grr.create_graphics_pipeline(grr::VertexPipelineDesc {
    vertex_shader: vs,
    tessellation_control_shader: None,
    tessellation_evaluation_shader: None,
    geometry_shader: None,
    fragment_shader: Some(fs),
})?;
```
Thoughts appreciated!

**Comment by msiglreith** [comment]
Found a reasonable compromise to keep changes to a minimum and support both approaches I guess, leaving it up to the user what they prefer regarding destruction. The approach would basically provide a transform between `DeviceHandle` <-> `Device`, where `DeviceHandle` impls deref for `Device`. Codewise it wouldn't change anything beside and additional conversion step on device creation into `DeviceHandle`, which allows to make it `Copy`.

```rust
use crate::Device;
use std::ops::Deref;

#[derive(Copy, Clone)]
pub struct DeviceHandle(*mut Device);

impl DeviceHandle {
    pub fn create(device: Device) -> Self {
        let device = Box::new(device);
        DeviceHandle(Box::into_raw(device))
    }

    pub fn destroy(self) {
        unsafe { Box::from_raw(self.0); }
    }
}

impl Deref for DeviceHandle {
    type Target = Device;

    fn deref(&self) -> &Self::Target {
        let device = unsafe { Box::from_raw(self.0) };
        Box::leak(device)
    }
}
```

As it would be quite small I'm not sure if it's worth moving it into the library, as others might want to tradeoff speed for safety via `Rc`/`Arc` for example.

**Comment by msiglreith** [comment]
Published 0.7 with the object handle changes including a few fixes and changes.
Leaving the Device stuff out, currently leaning towards letting the user decide how to store `Device`
